### PR TITLE
Add documentation for the KSQL docker images

### PR DESCRIPTION
**Pull request by apurvam**
This depends on:

1. https://github.com/confluentinc/ksql/pull/1322
2. https://github.com/confluentinc/ksql/pull/1347

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -45,8 +45,13 @@ The table below lists the available images and the Confluent software packages t
...
+| KSQL CLI         | cp-ksql-cli                  | Open Source  | - ksql-cli                              |
++------------------+------------------------------+--------------+-----------------------------------------+
 
-Note: The Kafka Connect image is labeled as "Enterprise" simply because it contains the Confluent Control Center package.  That package enables the deployed connectors to collect the metrics visualized in Confluent Control Center.   No explicit license is required when using the Kafka Connect image on its own.
+Note: The Kafka Connect image is labeled as "Enterprise" simply because it contains the Confluent Control Center package.  That package enables the deployed connectors to collect the metrics visualized in Confluent Control Center.   No explicit license is required when using the Kafka Connect image on its own. Similarly, the KSQL Server image is labelled "Enterprise" because it contains monitoring interceptors which enable the collection of metrics which can be visualized in Confluent Control Center. No explicit license is required to use the KSQL server on its own.
```
Comment (big-andy-coates) [inline docs/configuration.rst]: Maybe just say `Both the Kafka Connect and KQL Server images are labeled...`

File: docs/configuration.rst [hunk 2]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+The following settings must be passed to run the KSQL Server image.
+
+``KSQL_BOOTSTRAP_SERVERS``
+
+  A list of host/port pairs to use for establishing the initial connection to the Kafka cluster. The client will make use of all servers irrespective of which servers are specified here for bootstrapping; this list only impacts the initial hosts used to discover the full set of servers. This list should be in the form host1:port1,host2:port2,.... Since these servers are just used for the initial connection to discover the full cluster membership (which may change dynamically), this list need not contain the full set of servers (you may want more than one, though, in case a server is down).
+
+Optional Settings
+"""""""""""""""""
+All other settings for KSQL like security, monitoring interceptors, streams application overrides can be passed to the Docker images as environment variables. The names of these environment variables are derived by replacing ``.`` with ``_``, converting the resulting string to uppercase and prefixing it with ``KSQL_``. For example, if you need to set ``ksql.streams.state.dir``, the environment variable name would be ``KSQL_KSQL_STREAMS_STATE_DIR``.
```
Comment (big-andy-coates) [inline docs/configuration.rst]: Consider linking to the docs section covering security config here.

File: docs/configuration.rst [hunk 3]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+--------
+KSQL CLI
+--------
+The KSQL CLI image will execute the ``ksql`` command and try to connect to the server at ``http://localhost:8088`` by default, as shown below. 
+
+.. sourcecode:: bash
+
+  docker run -it -rm confluentinc/cp-ksql-cli:4.1.2
+
+Note: Since the CLI is interactive, we cannot log to stdout as is customary for Docker containers. The KSQL CLI logs can instead be found at ``/var/logs/ksql-cli/``.
```
Comment (big-andy-coates) [inline docs/configuration.rst]: > found at ``/var/logs/ksql-cli/``

Within the docker image, or the local host?

File: docs/configuration.rst [hunk 4]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+
+Configuring the KSQL CLI
+""""""""""""""""""""""""
+* Config files: The KSQL CLI config file must be passed in via a file mounted in the image.
```
Comment (big-andy-coates) [inline docs/configuration.rst]: Consider spelling it out a bit more, e.g. `Given that the CLI's property file is available on the host machine at '~/workspace/cli-config/ksql-cli.properties', then the following command will make the file available to the CLI running in the docker image:`

File: docs/configuration.rst [hunk 5]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+
+.. sourcecode:: bash
+  docker run -it -rm \
+    -v ~/workspace/cli-config:/mnt/ksql-cli
+    confluentinc/cp-ksql-cli:4.1.2 \
+      http://ksql-server.host:8088 \
+      --config-file /mnt/ksql-cli/ksql-cli.properties
+
+
+* Additional CLI options can be appended to the command as necessary.
```
Comment (miguno) [inline docs/configuration.rst]: Are there any? Can we give an example?

File: docs/configuration.rst (hunk 4 above)
Comment (miguno) [inline docs/configuration.rst]: Can't we configure the KSQL CLI also via environment variables?  The sections above imply Yes.

For example, we write (note the plural for images):
> All other settings for KSQL like security, monitoring interceptors, streams application overrides can be passed to the Docker images [...]

File: docs/configuration.rst [hunk 6]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+
+Note: Since the CLI is interactive, we cannot log to stdout as is customary for Docker containers. The KSQL CLI logs can instead be found at ``/var/logs/ksql-cli/``.
+
+Configuring the KSQL CLI
+""""""""""""""""""""""""
+* Config files: The KSQL CLI config file must be passed in via a file mounted in the image.
+
+.. sourcecode:: bash
+  docker run -it -rm \
+    -v ~/workspace/cli-config:/mnt/ksql-cli
```
Comment (miguno) [inline docs/configuration.rst]: Suggestion to clarify this for Docker beginners:

```
-v ~/workspace/cli-config:/mnt/ksql-cli
```
should be

```
-v /path/on/docker/host/to/cli-config:/path/in/container/ksql-cli
```

with a matching

```
--config-file /path/in/container/ksql-cli/ksql-cli.properties
```

File: docs/configuration.rst (hunk 4 above)
Comment (miguno) [reply docs/configuration.rst]: Yeah.  Taking my suggestion below into account, I'd rephrase slightly to:

> Assuming that the CLI's configuration file is available on the Docker host at `/path/on/docker/host/to/cli-config/ksql-cli.properties`, then the following command will make the configuration file available to the CLI container at `/path/in/container/ksql-cli/ksql-cli.properties`:

File: docs/configuration.rst [hunk 7]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+.. sourcecode:: bash
+
+  docker run -d \
+    -p 127.0.0.1:8088:8088
+    -e KSQL_BOOTSTRAP_SERVERS=localhost:9092 \
+    -e KSQL_LISTENERS=http://0.0.0.0:8088 \
+    -e KSQL_KSQL_SERVICE_ID=ksql_test_service_id_ \
+    -e KSQL_KSQL_STREAMS_REPLICATION_FACTOR=3 \
+    -e KSQL_KSQL_SINK_REPLICAS=3 \
+    -e KSQL_KSQL_STREAMS_STATE_DIR=/mnt/data/ksql-server-data \
```
Comment (miguno) [inline docs/configuration.rst]: We should clarify the paths for Docker beginners, this is so easy to get wrong (anecdotal evidence: while typing the suggestion below I first mixed up the paths, doh, even though I was focused on 'the path problem').

I'd suggest to change
```
-e KSQL_KSQL_STREAMS_STATE_DIR=/mnt/data/ksql-server-data \
-v /mnt/ksql-server/data:/mnt/data/ksql-server-data/ \
```

to

```
-e KSQL_KSQL_STREAMS_STATE_DIR=/path/in/container/ksql-server-data \
-v /path/on/docker/host/ksql-server-data:/path/in/container/ksql-server-data \
```

File: docs/configuration.rst [hunk 8]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+Configuring the KSQL CLI
+""""""""""""""""""""""""
+* Config files: The KSQL CLI config file must be passed in via a file mounted in the image.
+
+.. sourcecode:: bash
+  docker run -it -rm \
+    -v ~/workspace/cli-config:/mnt/ksql-cli
+    confluentinc/cp-ksql-cli:4.1.2 \
+      http://ksql-server.host:8088 \
+      --config-file /mnt/ksql-cli/ksql-cli.properties
```
Comment (miguno) [inline docs/configuration.rst]: Assuming you accept my suggestion above, this should become:
```
--config-file /path/in/container/ksql-cli/ksql-cli.properties
```

File: docs/configuration.rst (hunk 3 above)
Comment (apurvam) [reply docs/configuration.rst]: @miguno you can't use `docker log` for the CLI image. This is basically impossible, since the CLI uses stdout for real output. Directing the logs to stdout as well makes it unusable.

Comment (apurvam) [reply docs/configuration.rst]: (docker log only emits stdout and stderr from process id 0, which is the launch command of the container).

File: docs/configuration.rst (hunk 4 above)
Comment (apurvam) [reply docs/configuration.rst]: Good catch. We can't configure the CLI through the environment. Will update the wording.

File: docs/configuration.rst [hunk 9]
```diff
@@ -45,8 +45,13 @@ The table below lists the available images and the Confluent software packages t
 +------------------+------------------------------+--------------+-----------------------------------------+
 | REST Proxy       | cp-kafka-rest                | Open Source  | - confluent-kafka-rest                  |
 +------------------+------------------------------+--------------+-----------------------------------------+
+| KSQL Server      | cp-ksql-server               | Enterprise   | - ksql-rest-app                         |
```
[the remark this answers is not in the document]
Comment (apurvam) [reply docs/configuration.rst]: The nomenclature follows the rest. The kafka connect image includes interceptors, and his hence labelled 'entreprise'. I can put an asterisk next to the enterprise label and then link it to the note below that way.

I am not sure how the packages are defined elsewhere, so I just put in the module name for `ksql-rest-app`. I think its better to just call it `ksql-server`. Will update.

File: docs/configuration.rst [hunk 10]
```diff
@@ -472,3 +477,68 @@ Optional Settings
...
+The image will then convert these environment variables to corresponding KSQL Server config variables.
+
+--------
+KSQL CLI
+--------
+The KSQL CLI image will execute the ``ksql`` command and try to connect to the server at ``http://localhost:8088`` by default, as shown below. 
+
+.. sourcecode:: bash
+
+  docker run -it -rm confluentinc/cp-ksql-cli:4.1.2
```
[the remark this answers is not in the document]
Comment (apurvam) [reply docs/configuration.rst]: This is inconsistent with the other snippets.

File: docs/configuration.rst (hunk 4 above)
Comment (apurvam) [reply docs/configuration.rst]: Actually, the sentence you reference is for the KSQL server image. I don't think there is any implication that the CLI can be configured via environment variables. I will add a line to make that explicit.

**Comment by apurvam** [comment]
@miguno I addressed all your comments. Please take a look.

File: docs/configuration.rst [hunk 11]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+.. sourcecode:: bash
+
+  docker run -d \
+    -p 127.0.0.1:8088:8088
+    -e KSQL_BOOTSTRAP_SERVERS=localhost:9092 \
+    -e KSQL_LISTENERS=http://0.0.0.0:8088 \
+    -e KSQL_KSQL_SERVICE_ID=ksql_test_service_id_ \
+    -e KSQL_KSQL_STREAMS_REPLICATION_FACTOR=3 \
+    -e KSQL_KSQL_SINK_REPLICAS=3 \
+    -e KSQL_KSQL_STREAMS_STATE_DIR=/ \
```
Comment (rodesai) [inline docs/configuration.rst]: Should this be `/docker/container/path` so that the state stores are run on the mounted volume?

Comment (apurvam) [reply docs/configuration.rst]: good catch. fixed.

File: docs/configuration.rst [hunk 12]
```diff
@@ -493,27 +493,27 @@ The KSQL Server image uses variables prefixed with ``KSQL_`` with an underscore
...
 
 Optional Settings
 """""""""""""""""
-All other settings for KSQL like security, monitoring interceptors, streams application overrides can be passed to the Docker images as environment variables. The names of these environment variables are derived by replacing ``.`` with ``_``, converting the resulting string to uppercase and prefixing it with ``KSQL_``. For example, if you need to set ``ksql.streams.state.dir``, the environment variable name would be ``KSQL_KSQL_STREAMS_STATE_DIR``.
+All other settings for KSQL like :ref:`security <ksql-security>`, monitoring interceptors, overrides for Kakfa Streams, Kafka Producer, or Kakfa Consumer settings can be passed to the Docker images as environment variables. The names of these environment variables are derived by prefixing with KSQL_, uppercasing each word in the parameter name, and separating each word with an underscore (_) instead of periods (.). For example, if you need to set ``ksql.streams.state.dir``, the environment variable name would be ``KSQL_KSQL_STREAMS_STATE_DIR``.
```
Comment (miguno) [inline docs/configuration.rst]: - Typo: "Kakfa Streams" -> "Kafka Streams" ;-)
- I'd keep the backticks around ``KSQL_`` and `_` and `.` in this paragraph, just like you originally wrote.

File: docs/configuration.rst [hunk 13]
```diff
@@ -493,27 +493,27 @@ The KSQL Server image uses variables prefixed with ``KSQL_`` with an underscore
...
+To configure the Kafka cluster that KSQL should read from/write to, you must specify one or more "bootstrap" brokers from that Kafka cluster. This parameter is formatted as a comma-separated list of hostname:port pairs; for example, broker-hostname1:9092,broker-hostname2:9092. These brokers are used by KSQL to establish the initial connection to the Kafka cluster and to discover the set of all available brokers in the Kafka cluster, the set of which may change dynamically at runtime (e.g. brokers may be taken down for maintenance). You can but don't need to specify all brokers in the Kafka cluster, though it is recommended to specify at least two brokers to make bootstrapping more resilient against individual broker outages.
```
Comment (miguno) [inline docs/configuration.rst]: I'd suggest to use backticks around `broker-hostname1:9092,broker-hostname2:9092` to make it clear what of the paragraph text is the config value.

File: docs/configuration.rst [hunk 14]
```diff
@@ -526,19 +526,20 @@ The KSQL CLI image will execute the ``ksql`` command and try to connect to the s
 
   docker run -it -rm confluentinc/cp-ksql-cli:4.1.2
 
-Note: Since the CLI is interactive, we cannot log to stdout as is customary for Docker containers. The KSQL CLI logs can instead be found at ``/var/logs/ksql-cli/``.
+Note: Because the CLI is interactive, the container cannot log to STDOUT as is customary for Docker containers. The KSQL CLI logs can instead be found at /var/logs/ksql-cli/ inside the container.
```
Comment (miguno) [inline docs/configuration.rst]: I'd retain the backticks around the path, i.e. it should be ``/var/logs/ksql-cli/``

File: docs/configuration.rst [hunk 15]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+    -e KSQL_KSQL_SINK_REPLICAS=3 \
+    -e KSQL_KSQL_STREAMS_STATE_DIR=/docker/container/path \
+    -v /docker/host/path:/docker/container/path \
+    confluentinc/cp-ksql-server:4.1.2
+
+Docker Options
+""""""""""""""
+
+* Data persistence: KSQL Server runs Kafka Streams applications which may need to store persistent state locally. The command above maps the state store directory inside the container (cf. KSQL_KSQL_STREAMS_STATE_DIR) to a directory on the Docker host.
+* Port mapping: The above command maps the listener port to 8088 on the Docker host, hence ensuring that other containers (like for the KSQL CLI) running on the same host can connect to the KSQL server container.
```
Comment (miguno) [inline docs/configuration.rst]: Nit: We should be consistent when saying "KSQL Server" vs. "KSQL server"?

File: docs/configuration.rst [hunk 16]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+
+Required Settings
+"""""""""""""""""
+The following settings must be passed to run the KSQL Server image.
+
+``KSQL_BOOTSTRAP_SERVERS``
+
+To configure the Kafka cluster that KSQL should read from/write to, you must specify one or more "bootstrap" brokers from that Kafka cluster. This parameter is formatted as a comma-separated list of hostname:port pairs; for example, broker-hostname1:9092,broker-hostname2:9092. These brokers are used by KSQL to establish the initial connection to the Kafka cluster and to discover the set of all available brokers in the Kafka cluster, the set of which may change dynamically at runtime (e.g. brokers may be taken down for maintenance). You can but don't need to specify all brokers in the Kafka cluster, though it is recommended to specify at least two brokers to make bootstrapping more resilient against individual broker outages.
```
Comment (miguno) [inline docs/configuration.rst]: (this comment was IIRC lost/hidden because it was on a prior PR commit)

We should use backticks for ``broker-hostname1:9092,broker-hostname2:9092`` to make it extra clear which part of the paragraph text is the config value.  It also easier to find the example in the paragraph for people that read quickly.

File: docs/configuration.rst [hunk 17]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+
+Optional Settings
+"""""""""""""""""
+All other settings for KSQL like :ref:`security <ksql-security>`, monitoring interceptors, overrides for Kakfa Streams, Kafka Producer, or Kakfa Consumer settings can be passed to the Docker images as environment variables. The names of these environment variables are derived by prefixing with KSQL_, uppercasing each word in the parameter name, and separating each word with an underscore (_) instead of periods (.). For example, if you need to set ``ksql.streams.state.dir``, the environment variable name would be ``KSQL_KSQL_STREAMS_STATE_DIR``.
```
Comment (miguno) [inline docs/configuration.rst]: As you did in the original text, I'd prefer backticks around `KSQL_`, (`_`), and (`.`).

File: docs/configuration.rst [hunk 18]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+--------
+KSQL CLI
+--------
+The KSQL CLI image will execute the ``ksql`` command and try to connect to the server at ``http://localhost:8088`` by default, as shown below. 
+
+.. sourcecode:: bash
+
+  docker run -it -rm confluentinc/cp-ksql-cli:4.1.2
+
+Note: Because the CLI is interactive, the container cannot log to STDOUT as is customary for Docker containers. The KSQL CLI logs can instead be found at /var/logs/ksql-cli/ inside the container.
```
Comment (miguno) [inline docs/configuration.rst]: Can one still use `docker log ...` to see the STDOUT logs?

File: docs/configuration.rst [hunk 19]
```diff
@@ -472,3 +477,69 @@ Optional Settings
...
+.. sourcecode:: bash
+  docker run -it -rm \
+    -v /docker/host/path/:/docker/container/path
+    confluentinc/cp-ksql-cli:4.1.2 \
+      http://ksql-server.host:8088 \
+      --config-file /docker/container/path/ksql-cli.properties
+
+
+* Additional CLI options can be appended to the command as necessary. See :ref:`the KSQL CLI documentation <install_ksql-cli>` for the KSQL CLI options you can use.
+* The KSQL CLI cannot be configured via environment variables, unlike the KSQL Server image.
```
Comment (miguno) [inline docs/configuration.rst]: Why don't we allow configuring the CLI image via env variables?

**Comment by apurvam** [comment]
@miguno a note on timing. If this isn’t merged today there will be no docs for the docker images for the preview

File: docs/configuration.rst (hunk 19 above)
Comment (apurvam) [reply docs/configuration.rst]: It's based on my understanding of the way docker operates. It allows only one command to execute at the end. For other images, that is a script, which reads the environment, creates config files, and then runs the actual command.

In this case, the command is our `ksql` cli command. That doesn't leave any space for setting up config files. 

I may, however, be wrong about this limitation. I couldn't get it to work, but that doesn't mean it can't. We can fix it in a future version of the image, though.

File: docs/configuration.rst (hunk 18 above)
Comment (apurvam) [reply docs/configuration.rst]: See my response here: https://github.com/confluentinc/cp-docker-images/pull/496#discussion_r191879567

**Comment by apurvam** [comment]
Merging due to the time pressures (@miguno +1'd offline).
